linera-core/src/client.rs

```diff
@@ -1294,17 +1316,215 @@ where
+            Ok(_) => {
+                tracing::warn!("Ignoring invalid response from validator");
+                // Give up on this validator.
+                return Ok(());
+            }
+            Err(err) => {
+                tracing::warn!("Ignoring error from validator: {}", err);
+                return Ok(());
+            }
+        };
+
+        let certificates = future::try_join_all(
+            info.requested_sent_certificate_hashes
+                .into_iter()
+                .map(|hash| {
+                    let mut node = node.clone();
+                    async move { node.download_certificate(hash).await }
+                }),
+        )
+        .await?;
+
+        if !certificates.is_empty()
+            && self
+                .client
+                .local_node
+                .try_process_certificates(name, &mut node, chain_id, certificates, notifications)
+                .await
+                .is_none()
+        {
+            return Ok(());
+        };
+        if let Some(proposal) = info.manager.requested_proposed {
+            if proposal.content.block.chain_id == chain_id {
+                let owner = proposal.owner;
+                loop {
+                    let result = self
+                        .client
+                        .local_node
+                        .handle_block_proposal(*proposal.clone())
+                        .await;
+                    if let Err(original_err) = &result {
+                        if let LocalNodeError::WorkerError(WorkerError::ChainError(chain_error)) =
+                            original_err
+                        {
+                            if let ChainError::ExecutionError(
+                                ExecutionError::SystemError(
+                                    SystemExecutionError::BlobNotFoundOnRead(blob_id),
+                                ),
+                                _,
+                            ) = &**chain_error
+                            {
+                                self.update_local_node_with_blob(*blob_id).await?;
+                                continue;
+                            }
+                        }
+                        tracing::warn!("Skipping proposal from {}: {}", owner, original_err);
+                    }
+                    break;
+                }
+            }
+        }
+        if let Some(cert) = info.manager.requested_locked {
+            if cert.value().is_validated() && cert.value().chain_id() == chain_id {
+                let hash = cert.hash();
+                if let Err(error) = self
+                    .client
+                    .local_node
+                    .handle_certificate(*cert.clone(), vec![], vec![], notifications)
+                    .await
+                {
+                    tracing::warn!("Skipping certificate {}: {}", hash, error);
+                    return Err(error.into());
```

Comment:
If we get a locked block and are missing blobs we also need to download them. However, it suffices to get the blobs themselves, without the proving certificates.

Reply:
The validator might not have signed the original `ValidatedBlock` (e.g. because at the time it had a different locked block, or signed a conflicting proposal in the same round). But when they were given the `ValidatedBlock` _certificate_ they signed the `ConfirmedBlock` vote, and thus locked that block.

Reply:
Let me try that